### DISCUSSION TITLE
Provide service API keys or similar to end users

### DISCUSSION THREAD

**Discussion by mbalslow** [supabase/supabase#4419] [Questions] [answered]
This is more of a conceptual question on best practice approach when using Supabase.

In short:
Our clients need to also be able to access/interact with the data in our Supabase db programmatically (their own scripts, etc.).

Story:
We have made a web application. Works great, our end users can login using either email or social providers, and they can interact with the data in the frontend of the application with row level security to control access.

Now, we also need to be able to allow some (not all, only particular) users/accounts to interact with the data programmatically.

It's relevant for several use cases but to mention a specific one, we want to enable users to connect their BI platform to extract certain data to use in their reporting tools.

We don't want them to interact directly with the Supabase API, so it would be our own layer in-between, which in itself is not a problem.

The question/issue is that how should one generate/provide a long-living (or never expiring) API key/token, or a similar "static" way for people to authenticate, that is linked to a Supabase user so that when a request comes in, the [setAuth()](https://supabase.com/docs/reference/javascript/auth-setauth) method could be used to handle the request properly (respecting row level security).

Is there any way this can be done? Alternatively, one could simply set up any type of standard (Express or whatever) API with a completely separate authentication methodology and then for acceptable requests simply query Supabase using the secret key. But that would then completely ignore the row level security.

EDIT: Another approach we have considered was to allow for a /login endpoint in our custom API for username and password, which would simply perform a login via the SDK and the return the JWT for the end user to then use in the subsequent requests. But it just feels a bit unnatural. If that's the recommended approach then we could do that.

**Comment by steve-chavez** [comment]
> It's relevant for several use cases but to mention a specific one, we want to enable users to connect their BI platform to extract certain data to use in their reporting tools.

Seems you'd need a direct PostgreSQL connection for this case.

> But that would then completely ignore the row level security.
> Now, we also need to be able to allow some (not all, only particular) users/accounts to interact with the data programmatically.

If there are only a few of these users, then you could create a postgres role for each BI user(`create role bi_user_1`) and apply RLS using `current_user` instead of Supabase  `auth` functions. An example is show [here](https://postgrest.org/en/v9.0/auth.html#roles-for-each-web-user).

**Reply by mbalslow** [reply]
@steve-chavez it would not be a few users. It would be many. The BI example may be a bad one. Maybe a better example is like Stripe and Shopify API keys. We want to be able to allow end users to generate permanent/long-living API keys to enable themselves to access and interact (also create, update and delete) data programmatically. Potentially even allow users to manage what these keys will enable them of access.

**Reply by mbalslow** [reply]
Essentially, we are "simply" looking for a way to generate a permament/long-living JWT associated to a specific user serverside.

Right now, the only way we know of to generate a JWT, which would not even be permanent/long-living, is to use the signIn() method.

**Reply by mstade** [reply]
If you're comfortable with using the JWT secret in you API layer, then you can simply create your own JWT tokens there and supabase will accept them. Check out https://jwt.io/libraries for links to some useful libraries available. You may also want to follow #3899 and supabase/gotrue#277.

**Reply by mstade** [reply]
Also #1849 has some useful information specific to creating your own JWT tokens.

**Comment by FelixZY** [comment]
**EDIT**

I spent some time and wrote my own implementation since this does not appear supported yet. Feel free to take a look: https://gist.github.com/FelixZY/0aef530690458b381b8100afa19202c8

---

Would [this](https://github.com/supabase/supabase/issues/7186#issuecomment-1152992632) solve your issue?

**Reply by mbalslow** [reply]
Skimmed your described solution. Will definitely take it for a spin and get back to you.

**Reply by bepitulaz** [reply]
@FelixZY thank you for the solution. I also have similar use case. I'll try your solution.

**Reply by FelixZY** [reply]
Just making some noise to let you know I've released a significantly improved version of my originally suggested solution @mbalslow 
 @bepitulaz :slightly_smiling_face:

**Comment by vibegui** [comment]
Hi, just to weigh in, I'm also trying to create a token for an API to authenticate on Supabase and so far, I have found no strategy for that. Not sure I'm missing something obvious, but it feels super strange not being able to have an API Key. There are "access tokens" you can generate in the supabase admin, but I don't know to what end — couldn't find any way to use them.

**Comment by riderx** [comment, marked as answer]
I made this simple for my business [Capgo](https://capgo.app/).

If my RLS:
```sql
is_allowed_apikey(((current_setting('request.headers'::text, true))::json ->> 'capgkey'::text), '{all,write}'::key_mode[], app_id)
```
My table apikey:
| id | key (uuid)                           | user_id (uuid)                       | mode (all/write/read) |
|----|--------------------------------------|--------------------------------------|-----------------------|
| 1  | f83fd102-c21d-4984-b6a1-33c2cf108fd7 | f83fd102-c21d-4984-b6a1-33c2cf108fd7 | all                   |
my postgress function
```sql
CREATE OR REPLACE FUNCTION public.is_allowed_apikey(apikey text, keymode key_mode[])
 RETURNS boolean
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$
Begin
  RETURN (SELECT EXISTS (SELECT 1
  FROM apikeys
  WHERE key=apikey
  AND mode=ANY(keymode)));
End;  
$function$
```
And in the SDK 1 you can add your APIKEY like that
```javascript
const supabase = createClient(hostSupa, supaAnon, {
    headers: {
        capgkey: apikey,
    }
})
```
In SDK v2
```javascript
const supabase = createClient(hostSupa, supaAnon, {
    global: {
      headers: {
          capgkey: apikey,
      }
  }
})
```

I explained first this way there: https://github.com/supabase/storage-api/issues/81#issuecomment-1228507294

**Reply by vickkhera** [reply]
This is brilliant. Having a way to pass data from the client to the DB like this makes a bunch of things possible.

**Reply by fvermaut** [reply]
That looks nice!  
I need to try this out, but in the meantime I would have a question:  can you implement multi-tenant RLS policies with such approach?  
For instance, restricting access to users authenticating with a specific apikey to specific rows (based on user_id set on those rows, or any other identifier).  Thanks
